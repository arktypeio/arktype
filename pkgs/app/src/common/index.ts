import { BrowserName } from "@re-do/test"
import { Step, Tag, Element, Test } from "@re-do/model"
import type { ShallowModel } from "persist-statelessly"

export type RedoData = {
    tests: Test[]
    elements: Element[]
    steps: Step[]
    tags: Tag[]
}

export type Page = "HOME" | "SIGN_IN" | "SIGN_UP"

export type UnsavedStep = Step & { id: number }

export type Root = {
    page: Page
    token: string
    cardFilter: string
    defaultBrowser: BrowserName
    builder: {
        active: boolean
        steps: UnsavedStep[]
    }
    main: MainActions
    renderer: RendererActions
    data: ShallowModel<RedoData, "id">
}

export type MainActions = {
    runTest: [number] | null
    saveTest: [Test] | null
    launchBuilder: [] | null
    closeBuilder: [] | null
}

export type RendererActions = {}

export const supportedEvents = {
    click: "click",
    dblclick: "click",
    submit: "click",
    change: "set",
    select: "set"
} as const

export type SupportedEvent = keyof typeof supportedEvents
