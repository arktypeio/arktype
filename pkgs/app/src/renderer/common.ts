import {
    ApolloClient,
    InMemoryCache,
    createHttpLink,
    ApolloLink
} from "@apollo/client"
import { rootHandler, initialRoot } from "state"
import { createStore } from "react-statelessly"
import { isDev } from "@re-do/utils/dist/node"

const httpLink = createHttpLink({
    uri: isDev()
        ? `http://localhost:${process.env.PORT}/dev/graphql`
        : "https://tpru7v18yi.execute-api.us-east-1.amazonaws.com/dev/graphql"
})

const contextLink = new ApolloLink((operation, forward) => {
    operation.setContext({
        headers: {
            authorization: `Bearer ${store.query({ token: true }).token}`
        }
    })
    return forward(operation)
})

export const client = new ApolloClient({
    link: contextLink.concat(httpLink),
    cache: new InMemoryCache()
})
export const store = createStore({
    initial: initialRoot,
    handler: rootHandler
})
