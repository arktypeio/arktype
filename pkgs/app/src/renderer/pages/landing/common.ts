export const formatEmail = (email: string | undefined) =>
    email ? email.trim().toLowerCase() : ""
