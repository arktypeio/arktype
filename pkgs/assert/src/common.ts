import { Func } from "@re-/tools"
import { resolve } from "deno/std/path/mod.ts"

export type LinePosition = {
    line: number
    char: number
}

export type SourcePosition = LinePosition & {
    file: string
    method: string
}

export const writeJsonSync = (path: string, data: unknown) => {
    Deno.writeTextFileSync(path, JSON.stringify(data, null, 4))
}

export const readJsonSync = (path: string) => {
    try {
        return JSON.parse(Deno.readTextFileSync(path))
    } catch {
        return undefined
    }
}

export const setJsonKey = (path: string, key: string, value: unknown) =>
    writeJsonSync(path, { ...readJsonSync(path), [key]: value })

export const existsSync = (path: string) => {
    try {
        Deno.openSync(path).close()
        return true
    } catch {
        return false
    }
}

export interface ReAssertConfig extends Required<ReAssertJson> {
    updateSnapshots: boolean
}

export interface ReAssertJson {
    tsconfig?: string
    precached?: boolean
    precachePath?: string
}

export interface ReJson {
    assert?: ReAssertJson
}

export type Memoized<F extends Func> = F & { cache?: ReturnType<F> }

export const getReAssertConfig: Memoized<() => ReAssertConfig> = () => {
    if (!getReAssertConfig.cache) {
        const reJson: ReJson = readJsonSync("re.json") ?? {}
        const reAssertJson: ReAssertJson = reJson.assert ?? {}
        getReAssertConfig.cache = {
            updateSnapshots: !!Deno.args.find(
                (arg) => arg === "--update" || arg === "-u"
            ),
            tsconfig: resolve("tsconfig.json"),
            precached: false,
            precachePath: resolve(".assert.cache.json"),
            ...reAssertJson
        }
    }
    return getReAssertConfig.cache
}
