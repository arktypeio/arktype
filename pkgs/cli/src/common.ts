import { ensureDirSync } from "fs-extra"
import { homedir } from "os"
import { join } from "path"

export const redoDir = join(homedir(), ".redo")
ensureDirSync(redoDir)

const executableSuffixes = {
    darwin: "-mac.zip",
    linux: ".AppImage",
    win32: ".exe"
}

export type supportedOs = keyof typeof executableSuffixes

export const os = process.platform
if (!(os in executableSuffixes)) {
    throw new Error(`Redo does not support os: '${os}'.`)
}
export const executableSuffix = executableSuffixes[os as supportedOs]
export const appExecutable = join(redoDir, `redo${executableSuffix}`)
