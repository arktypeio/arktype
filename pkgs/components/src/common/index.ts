export const makeKinds =
    <Props>() =>
    <Kind extends string, Options = undefined>(
        kinds:
            | Record<Kind, Partial<Props>>
            | ((options: Options) => Record<Kind, Partial<Props>>)
    ) =>
        (typeof kinds === "function"
            ? (kind: Kind, options: Options) => kinds(options)[kind]
            : (kind: Kind) => kinds[kind]) as any as Options extends undefined
            ? (kind: Kind) => Partial<Props>
            : (kind: Kind, options: Options) => Partial<Props>

export type KindFrom<T extends (...args: any[]) => any> = Parameters<T>[0]
