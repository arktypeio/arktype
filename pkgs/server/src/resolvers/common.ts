import Photon, { FindOneUserArgs } from "@generated/photon"
import { TagInput } from "@re-do/model"
import { Context } from "../context"

type FindUserOptions = {
    photon: Photon
    query: FindOneUserArgs
}

// TODO: Find an idiomatic way to check if a unique field exists in photon
export const findUser = async ({ query, photon }: FindUserOptions) => {
    try {
        return await photon.users.findOne(query)
    } catch {
        return null
    }
}

type SplitResult<T> = [T[], T[]]

const split = <T>(list: T[], by: (item: T) => boolean) =>
    list.reduce(
        (sorted, item) =>
            (by(item)
                ? [[...sorted[0], item], sorted[1]]
                : [sorted[0], [...sorted[1], item]]) as SplitResult<T>,
        [[], []] as SplitResult<T>
    )

export const createTagConnector = async (
    tags: TagInput[],
    { photon, id }: Context
) => {
    const existing = (await photon.tags.findMany({
        where: { user: { id } },
        select: {
            name: true
        }
    })).map(_ => _.name)
    const [used, unused] = split(tags, ({ name }) => existing.includes(name))
    for (const { name } of unused) {
        await photon.tags.create({ data: { name, user: { connect: { id } } } })
    }
    return { connect: tags.map(({ name }) => ({ name })) }
}
