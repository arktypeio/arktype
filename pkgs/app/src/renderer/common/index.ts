import {
    ApolloClient,
    InMemoryCache,
    createHttpLink,
    ApolloLink
} from "@apollo/client"
import { stateSyncEnhancer } from "electron-redux/renderer"
import { Root } from "common"
import { Store } from "react-statelessly"

const uri = import.meta.env.DEV
    ? `http://localhost:${import.meta.env.VITE_GRAPHQL_SERVER_PORT}/dev/graphql`
    : "https://tpru7v18yi.execute-api.us-east-1.amazonaws.com/dev/graphql"

const httpLink = createHttpLink({ uri })

const contextLink = new ApolloLink((operation, forward) => {
    operation.setContext({
        headers: {
            authorization: `Bearer ${store.get("token")}`
        }
    })
    return forward(operation)
})

export const client = new ApolloClient({
    link: contextLink.concat(httpLink),
    cache: new InMemoryCache()
})

export const store = new Store(
    {} as Root,
    {},
    {
        reduxOptions: {
            enhancers: (enhancers) => [stateSyncEnhancer()].concat(enhancers)
        }
    }
)

// Test hook
store.update({ main: { __rendererLaunched: [] } })
