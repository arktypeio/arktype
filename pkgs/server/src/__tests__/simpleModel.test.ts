import { generateModel } from "../generateModel"
import { resolvers } from "./common"
import { join} from "path"
import { readFileSync, mkdtempSync, removeSync } from "fs-extra"

let inDir: string
let atPath: string

beforeAll(() => {
    inDir = mkdtempSync("temp-")
    atPath = join(inDir, "datamodel.prisma")
})

afterAll(() => {
    removeSync(inDir)
})

afterEach(() => {
    removeSync(atPath)
})

test("handles simple model", () => {
    generateModel({
        resolvers,
        atPath
    })
    expect(readFileSync(atPath).toString()).toMatchSnapshot()
})
