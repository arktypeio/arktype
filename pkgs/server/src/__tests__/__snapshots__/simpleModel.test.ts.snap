// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles simple model 1`] = `
"type A {
  a: [String!]!
}

type Fake {
  a: A!
  b: Boolean!
}

input FakeInput {
  first: String!
  second: Boolean!
}
"
`;
